ray-on-gke/kuberay-tpu-webhook/main.go

```diff
@@ -88,74 +93,192 @@ func extractRayCluster(admissionReview *admissionv1.AdmissionReview) (*ray.RayCl
+			envPath := fmt.Sprintf("/spec/workerGroupSpecs/%d/template/spec/containers/%d/env", workerGroupIndex, j)
 			tpuWorkerHostNames := corev1.EnvVar{
 				Name:  "TPU_WORKER_HOSTNAMES",
 				Value: hostNames,
 			}
+			subdomainPatch["path"] = subdomainPath
+			subdomainPatch["value"] = "tpu-worker-group-svc"
 			if len(container.Env) == 0 {
-				hostNamesPatch["path"] = path
+				hostNamesPatch["path"] = envPath
 				hostNamesPatch["value"] = []corev1.EnvVar{tpuWorkerHostNames}
 			} else {
-				hostNamesPatch["path"] = fmt.Sprintf("%s/-", path)
+				hostNamesPatch["path"] = fmt.Sprintf("%s/-", envPath)
 				hostNamesPatch["value"] = tpuWorkerHostNames
 			}
-			*patches = append(*patches, hostNamesPatch)
+			*patches = append(*patches, subdomainPatch, hostNamesPatch)
+		}
+	}
+}
+
+func checkWorkersMatchTopology(workerGroupSpec ray.WorkerGroupSpec) (bool, error) {
+	replicas := workerGroupSpec.Replicas
+	if replicas == nil {
+		return false, errors.New("workerGroupSpec replicas not set")
+	}
+	numWorkers := int(*replicas)
+	containers := workerGroupSpec.Template.Spec.Containers
+	if containers == nil {
+		return false, errors.New("Container path not specified")
+	}
+	if containerRequestingTPUs(containers...) {
+		topology := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
+		if topology == "" {
+			klog.Error("TPU topology not specified")
+		}
+		hosts, err := getNumTPUHosts(topology)
+		if err != nil {
+			return false, err
+		}
+
+		if hosts != numWorkers {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func isScheduledWithAffinity(workerGroupSpec ray.WorkerGroupSpec) (bool, error) {
+	containers := workerGroupSpec.Template.Spec.Containers
+	if containers == nil {
+		return false, errors.New("Container path not specified")
+	}
+	if containerRequestingTPUs(containers...) {
+		topology := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-tpu-topology"]
+		isMultiHost, err := isTPUMultiHost(topology)
+		if err != nil {
+			return false, err
+		}
+		if isMultiHost {
+			placementGroup := workerGroupSpec.Template.Spec.NodeSelector["cloud.google.com/gke-placement-group"]
+			if placementGroup == "" {
+				return false, nil	
+			}
+		}
+	}
+
+	return true, nil
+}
+
+func validateRayCluster(admissionReview *admissionv1.AdmissionReview) (*admissionv1.AdmissionResponse, error) {
+	raycluster, err := extractRayCluster(admissionReview)
+	if err != nil {
+		klog.Fatalf("Ray Cluster extraction failed: %s", err)
+	}
+
+	admit := true
+	status := "Success"
+	message := ""
+	workerGroupSpecs := raycluster.Spec.WorkerGroupSpecs
+	if workerGroupSpecs == nil {
+		return nil, errors.New("WorkerGroupSpecs not specified")
+	}
+	for i := 0; i < len(workerGroupSpecs); i++ {
+		workerGroupSpec := workerGroupSpecs[i]
+		workersMatchTopology, err := checkWorkersMatchTopology(workerGroupSpec)
+		if err != nil {
+			return nil, err
+		}
+		workerGroupAffinity, err := isScheduledWithAffinity(workerGroupSpec)
```

Comment:
`workerGroupAffinity` sounds like an object name, not a boolean name, maybe it change it something like `hasCorrectAffinity` or something.

Reply:
Done